docs/en/pact-reference.md

```diff
@@ -1277,9 +1277,200 @@ Declaring models shares the same syntax with modules:
+the object-oriented connotation of "objects changing shape". Modrefs allow modules to
+"interoperate with each other".
+
+Additionally, programmers should resist the urge to employ other object-oriented
+patterns using modrefs. A popular idea of "coding around interfaces, not
+implementations" is actually harmful in Pact if polymorphism is not
+needed. Programmers should prefer direct references whenever possible as they
+are not only faster but safer, as Pact aggressively and permanently inlines all
+directly-referenced code in the interests of code stability.
+
+### Important concerns when using modrefs.
+
+#### Late Binding
+
+As of Pact 4.7, modrefs are "late-binding", which means that the latest
+upgraded version of a module will be used when a module operation is invoked.
+
+Consider a modref to a module stored in the database when the module is
+at version 1. Sometime later the module is upgraded to version 2. The modref
+in the database will refer to the upgraded version 2 of the module when read
+back in and used.
+
+As described in the [Dependency Management](#dependency-management) section,
+Pact direct references are not late-binding, so this modref behavior might
+be surprising. "Pinned modrefs" that bind eagerly to the module version is
+being actively considered as an upcoming Pact change to extend the benefits of
+dependency management to modrefs.
+
+#### Modrefs introduce untrusted code
+
+As modrefs introduce the ability to call code unknown at compile time, this
+means that **any modref call should be treated as untrusted code**. Thus
+programmers should avoid invoking modrefs when a sensitive capability is in
+scope, as it will make "private capability functions" callable by the modref,
+when normally this would not be possible.
+
+Consider a module with a public function `collect-data` that is safe to call,
+and a private function `pay-fee` that is not intended to be invoked outside the
+module, guarded by the capability `COLLECT`. If a modref is invoked while the
+`COLLECT` capability is in scope, the calling code will be able to call the
+sensitive function directly.
+
+```lisp
+;; BAD CODE -- don't do this!
+(defun collect-data (collector:module{data-collector} account:string)
+  (with-capability (COLLECT)
+    ;; BAD: modref invoked inside of with-capability call
+    (store-data (collector::collect))
+    (pay-fee account)))
+
+(defun pay-fee (account:string)
+  (require-capability (COLLECT))
+  (coin.transfer FEE_BANK account FEE))
+```
+
+In the above code, `pay-fee` is intended to run once. But because the
+`collector:module{data-collector}` modref is invoked **inside of the capability
+acquisition**, the modref code in `collect` now has elevated privilege to
+directly call `pay-fee`. Thus a malicious actor can craft a version of `collect`
+thatn can directly call `pay-fee` as many times as they like.
+
+Fortunately, this is easily avoided by keeping modref calls out of scope of the
+sensitive capability.
+
```

Comment:
Is it not cheaper and easier to understand if we would just tell them to execute the `store-data` outside of the `with-capability` as well?

```pact
;; SAFE CODE -- do this!
(defun collect-data (collector:module{data-collector} account:string)
  ;; GOOD: modref invoked outside of with-capability call
  (store-data (collector::collect))
  (with-capability (COLLECT)
    (pay-fee account))))
```

Or does invoking it like this have different implications when the capability cannot be acquired?

Reply:
It's a made-up example, I just wanted to show the code doing _something_ with the results of the collector call. Presumably `store-data` is a sensitive call too (you wouldn't want someone making that call top-level). But this is the problem with throwing examples out without a massive section detailing them. I'm rewriting the intro to this so I'll try to clarify.